tests/components/icloud/test_config_flow.py

```diff
@@ -0,0 +1,279 @@
+from unittest.mock import patch, MagicMock
+import pytest
+
+from pyicloud.exceptions import PyiCloudFailedLoginException
+
+from homeassistant import data_entry_flow
+from homeassistant.components.icloud import config_flow
+
+from homeassistant.components.icloud.config_flow import (
+    CONF_TRUSTED_DEVICE,
+    CONF_VERIFICATION_CODE,
+)
+from homeassistant.components.icloud.const import (
+    DOMAIN,
+    CONF_ACCOUNT_NAME,
+    CONF_GPS_ACCURACY_THRESHOLD,
+    CONF_MAX_INTERVAL,
+    DEFAULT_GPS_ACCURACY_THRESHOLD,
+    DEFAULT_MAX_INTERVAL,
+)
+from homeassistant.const import CONF_PASSWORD, CONF_USERNAME
+from homeassistant.helpers.typing import HomeAssistantType
+
+from tests.common import MockConfigEntry
+
+USERNAME = "username@me.com"
+PASSWORD = "password"
+ACCOUNT_NAME = "Account name 1 2 3"
+ACCOUNT_NAME_FROM_USERNAME = None
+MAX_INTERVAL = 15
+GPS_ACCURACY_THRESHOLD = 250
+
+TRUSTED_DEVICES = [
+    {"deviceType": "SMS", "areaCode": "", "phoneNumber": "*******58", "deviceId": "1"}
+]
+
+
+@pytest.fixture(name="service")
+def mock_controller_service():
+    """Mock a successful service."""
+    with patch("pyicloud.base.PyiCloudSession"):
+        with patch("pyicloud.base.PyiCloudService", requires_2fa=True) as service_mock:
+            yield service_mock
+
+
+@pytest.fixture(name="service_with_cookie")
+def mock_controller_service_with_cookie():
+    """Mock a successful service while already authenticate."""
+    with patch("pyicloud.base.PyiCloudSession"):
+        with patch("pyicloud.base.PyiCloudService", requires_2fa=False) as service_mock:
+            yield service_mock
+
+
+@pytest.fixture(name="service_send_verification_code_failed")
+def mock_controller_service_send_verification_code_failed():
+    """Mock a failed service during sending verification code step."""
+    with patch("pyicloud.base.PyiCloudSession"):
+        with patch(
+            "pyicloud.base.PyiCloudService",
+            requires_2fa=False,
+            trusted_devices=TRUSTED_DEVICES,
+        ) as service_mock:
+            service_mock.send_verification_code.return_value = False
+            yield service_mock
+
+
+@pytest.fixture(name="service_validate_verification_code_failed")
+def mock_controller_service_validate_verification_code_failed():
+    """Mock a failed service during validation of verification code step."""
+    with patch("pyicloud.base.PyiCloudSession"):
+        with patch(
+            "pyicloud.base.PyiCloudService",
+            requires_2fa=False,
+            trusted_devices=TRUSTED_DEVICES,
+        ) as service_mock:
+            service_mock.send_verification_code.return_value = True
+            service_mock.validate_verification_code.return_value = False
+            yield service_mock
+
+
+def init_config_flow(hass: HomeAssistantType, service: MagicMock):
+    """Init a configuration flow."""
+    flow = config_flow.IcloudFlowHandler()
+    flow.hass = hass
+    flow.api = service
```

Comment:
Don't do this. Patching the class and setting the return value of the class will be enough. We shouldn't interfere with the code under test.